pkg/controllers/management/auth/globalroles/fleetpermissions/handler.go

```diff
@@ -0,0 +1,297 @@
+	GRBFleetWorkspaceOwnerLabel      = "authz.management.cattle.io/grb-fw-owner"
+	localFleetWorkspace              = "fleet-local"
+	fleetWorkspaceClusterRulesPrefix = "fwcr-"
+	fleetWorkspaceVerbsPrefix        = "fwv-"
+)
+
+type Handler struct {
+	crClient  rbacv1.ClusterRoleController
+	crCache   rbacv1.ClusterRoleCache
+	crbClient rbacv1.ClusterRoleBindingController
+	crbCache  rbacv1.ClusterRoleBindingCache
+	grCache   mgmtcontroller.GlobalRoleCache
+	rbClient  rbacv1.RoleBindingController
+	rbCache   rbacv1.RoleBindingCache
+	fwCache   mgmtcontroller.FleetWorkspaceCache
+}
+
+func NewHandler(management *config.ManagementContext) *Handler {
+	return &Handler{
+		crClient:  management.Wrangler.RBAC.ClusterRole(),
+		crCache:   management.Wrangler.RBAC.ClusterRole().Cache(),
+		crbClient: management.Wrangler.RBAC.ClusterRoleBinding(),
+		crbCache:  management.Wrangler.RBAC.ClusterRoleBinding().Cache(),
+		grCache:   management.Wrangler.Mgmt.GlobalRole().Cache(),
+		rbClient:  management.Wrangler.RBAC.RoleBinding(),
+		rbCache:   management.Wrangler.RBAC.RoleBinding().Cache(),
+		fwCache:   management.Wrangler.Mgmt.FleetWorkspace().Cache(),
+	}
+}
+
+// ReconcileFleetWorkspacePermissions creates or updates backing ClusterRoles and bindings for fleet workspace permissions.
+// Permissions specified in InheritedFleetWorkspacePermissions will apply to all fleet workspaces except local.
+// InheritedFleetWorkspacePermissions.ResourceRules applies to the namespace of the workspace.
+// InheritedFleetWorkspacePermissions.WorkspaceVerbs applies to the cluster-wide fleetworkspace resources.
+func (h *Handler) ReconcileFleetWorkspacePermissions(globalRoleBinding *v3.GlobalRoleBinding) error {
+	globalRole, err := h.grCache.Get(globalRoleBinding.GlobalRoleName)
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole: %w", err)
+	}
+	if globalRole.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil &&
+		globalRole.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return nil
+	}
+	fleetWorkspaces, err := h.fwCache.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list fleetWorkspaces when reconciling globalRoleBinding %s: %w", globalRoleBinding.Name, err)
+	}
+
+	err = h.reconcilePermissionsResourceRules(globalRoleBinding, globalRole, fleetWorkspaces)
+	if err != nil {
+		return fmt.Errorf("error reconciling fleet permissions rules: %w", err)
+	}
+
+	err = h.reconcilePermissionsWorkspaceVerbs(globalRoleBinding, globalRole, fleetWorkspaces)
+	if err != nil {
+		return fmt.Errorf("error reconciling fleet workspace verbs: %w", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) reconcilePermissionsResourceRules(globalRoleBinding *v3.GlobalRoleBinding, globalRole *v3.GlobalRole, fleetWorkspaces []*v3.FleetWorkspace) error {
+	fleetWorkspacePermissionName := fleetWorkspaceClusterRulesPrefix + globalRole.Name
+
+	err := h.reconcileRulesClusterRole(globalRole, globalRoleBinding.Name, fleetWorkspacePermissionName)
```

Comment:
The idea was to avoid creating the backing `ClusterRoles` until there were at least one `GlobalRoleBinding` to minimize the number of resources created. However, I agree with you now that It'd would be better to do it in the GlobalRole reconciler instead considering the reasons you mentioned. Additionally, this will also make easier the validation in the webhook. Changed in https://github.com/rancher/rancher/pull/44901/commits/f7974c82cfdaed74741e02d383e3e435b6f9a850